wgpu-core/src/device.rs

```diff
@@ -940,12 +940,19 @@ impl<F: IdentityFilter<BufferId>> Global<F> {
     pub fn buffer_destroy<B: GfxBackend>(&self, buffer_id: BufferId) {
         let hub = B::hub(self);
         let mut token = Token::root();
-        let (device_guard, mut token) = hub.devices.read(&mut token);
-        let (buffer_guard, _) = hub.buffers.read(&mut token);
-        let buffer = &buffer_guard[buffer_id];
-        device_guard[buffer.device_id.value].pending.lock().destroy(
+
+        // Deadlock hotfix, pending doesn't have a token
```

Comment:
This isn't a deadlock hotfix, if I understand correctly. As we've been discussing here, it's just sort of an optimization thing: to not hold the buffer storage locked for the duration of the `pending.lock()`

Reply:
No, the actual issue was a deadlock, not a performance issue. Rendering would completely stop and hang on these cleanup functions and in `triage_referenced` in `maintain`. This is because they lock in different orders.
`maintain` locks `pending` first, then the resources, these functions other way around.
Separating this so the cleanup functions don't need both locks at the same time resolves the deadlock.